packages/core-data/src/selectors.ts

```diff
@@ -215,14 +226,56 @@ export function getEntity( state: State, kind: string, name: string ): any {
  *
  * @return Entity config
  */
-export function getEntityConfig(
-	state: State,
-	kind: string,
-	name: string
-): any {
+export function getEntityConfig( state: State, kind: Kind, name: Name ): any {
 	return find( state.entities.config, { kind, name } );
 }
 
+/**
+ * GetEntityRecord is declared as an *interface*, but it actually describes
+ * the specifies the getEntityRecord *function* signature. It may seem unusual,
+ * but it's just how TypeScript implements function overloading.
+ *
+ * More accurately, GetEntityRecord distinguishes between two different signatures
+ * the getEntityRecord selector has:
+ *
+ * 1. When query._fields is not given, the returned type is EntityRecordOf< K, N, C >
+ * 2. When query._fields is given, the returned type is Partial<EntityRecordOf< K, N, C >>
+ *
+ * Unfortunately, due to a TypeScript limitation (https://github.com/microsoft/TypeScript/issues/23132)
+ * we can't use a single function signature with a return type such as:
+ *
+ *    Fields extends undefined
+ * 	    ? EntityRecordOf< K, N, C >
+ * 		  : Partial< EntityRecordOf< K, N, C > >
+ */
+interface GetEntityRecord {
+	<
+		R extends EntityRecordOf< K, N >,
+		C extends Context = DefaultContextOf< R >,
+		K extends Kind = KindOf< R >,
+		N extends Name = NameOf< R >
+	>(
+		state: State,
+		kind: K,
+		name: N,
+		key: KeyOf< K, N >,
+		query: EntityQuery< C >
+	): Partial< EntityRecordOf< K, N, C > > | null | undefined;
```

Comment:
It would be nice if it were a compile time error to refer to a field that you didn't select.

```ts
const stuff = core.getEntityRecords( 'root', 'comment', { _fields: [ 'id', 'content' ] } );
console.log( stuff[ 0 ].title ); // shouldn't compile
```

Maybe this is possible by having the user specify (again...) which fields they are selecting as a type param.

```ts
const stuff = core.getEntityRecords< 'id' | 'content' >( 'root', 'comment', { _fields: [ 'id', 'content' ] } );
console.log( stuff[ 0 ].title ); // shouldn't compile
```

Reply:
Yes it would! I hope to make it happen in one of the next iterations – let's keep this PR at atomic as possible.